draft-ietf-tls-esni.md

```diff
@@ -572,12 +574,57 @@ trigger retries, as described in {{handle-server-response}}. This may be
 implemented, for instance, by reporting a failed connection with a dedicated
 error code.
 
-### GREASE extensions {#grease-extensions}
+### HelloRetryRequest
+
+If the server sends a HelloRetryRequest in response to the ClientHello,
+the client sends a second updated ClientHello per the rules in {{RFC8446}}.
+However, at this point, the client does not know whether the server processed
+ClientHelloOuter or ClientHelloInner, and MUST regenerate both values to
+be acceptable. Note: if the inner and outer ClientHellos use different groups
+for their key shares or differ in some other way, then the HelloRetryRequest may
+actually be invalid for one or the other ClientHello. If the inner ClientHello
+is unaffected by this retry, then the client only changes the outer ClientHello.
+In contrast, if the outer ClientHello is unaffected by this retry, then the client
+changes the inner ClientHello and recomputes any fields necessary in the outer
+ClientHello ("encrypted_client_hello" extension contents, PSK binders, etc.)
+
+Otherwise, the usual rules for HelloRetryRequest processing apply.
+
+Clients bind encryption of the second ClientHelloInner to encryption of the first
+ClientHelloInner via the derived echo_hrr_key by modifying HPKE setup as follows:
+
+~~~
+pkR = HPKE.KEM.Unmarshal(ECHOConfig.public_key)
+enc, context = SetupPSKS(pkR, "tls13-echo-hrr", echo_hrr_key, "")
```

Comment:
See my previous comments about ctx rather than PSK. Maybe an open issue marker here?

Reply:
I went back and double checked the PSK variant of HPKE, and `info` is assumed to be known to the attacker in our analysis. Here that's not the case, so I *think* PSK is the right place for it. But agreed, let's drop a TODO to follow up with Karthik et al.